# Doctor and Veterinary Classification using NLP

This notebook is for building a model which will correctly classify a number of given reddit users as practicing doctors, practicng veterinary or others based on each user's comments 

The dataset for this task would be sourced from a databased whose link is given as

[postgresql://niphemi.oyewole:W7bHIgaN1ejh@ep-delicate-river-a5cq94ee-pooler.us-east-2.aws.neon.tech/Vetassist?statusColor=F8F8F8&env=&name=redditors%20db&tLSMode=0&usePrivateKey=false&safeModeLevel=0&advancedSafeModeLevel=0&driverVersion=0&lazyload=false](postgresql://niphemi.oyewole:W7bHIgaN1ejh@ep-delicate-river-a5cq94ee-pooler.us-east-2.aws.neon.tech/Vetassist?statusColor=F8F8F8&env=&name=redditors%20db&tLSMode=0&usePrivateKey=false&safeModeLevel=0&advancedSafeModeLevel=0&driverVersion=0&lazyload=false)

However, trying to access the database with the given link would result in errors

Therefore, a modified version of the link would be used

## Module Importations and Data Retrieval

Before continuing, needed libraries would be imported below

In [1]:
import re             # for regrex operations
import string         # for removing punctuations
import numpy as np    # for mathematical calculations
import pandas as pd    # for working with structured data (dataframes)
from sqlalchemy import create_engine # for connecting to database
from nltk.tokenize import word_tokenize
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Embedding
from tensorflow.keras.initializers import Constant
from sklearn.model_selection import train_test_split

In [2]:
from tensorflow.keras.preprocessing.text import Tokenizer
from autocorrect import Speller
from nltk.corpus import stopwords

In [81]:
import random

The modified link to access the database is defined below

In [3]:
# # define the connection link
# conn_str = "postgresql://niphemi.oyewole:endpoint=ep-delicate-river-a5cq94ee-pooler;W7bHIgaN1ejh@ep-delicate-river-a5cq94ee-pooler.us-east-2.aws.neon.tech/Vetassist?sslmode=allow"

# # create connection to the databse
# engine =  create_engine(conn_str)

First, lets take a look at the tables in the database

In [4]:
# define sql query for retrieving the tables in the database
sql_for_tables = """
SELECT
    table_schema || '.' || table_name
FROM
    information_schema.tables
WHERE
    table_type = 'BASE TABLE'
AND
    table_schema NOT IN ('pg_catalog', 'information_schema');
"""

In [5]:
# # retrieve the tables in a dataframe
# tables_df = pd.read_sql_query(sql_for_tables, engine)

In [6]:
# tables_df

There are two tables in the database as shown above

Each table would be saved in a pandas dataframe

In [7]:
sql_for_table1 = """
SELECT
    *
FROM
    public.reddit_usernames_comments;
"""

> Note: The code below may take a while to run. If it fails, reconnect the engine above then rerun the cell

In [8]:
# user_comment_df = pd.read_sql_query(sql_for_table1, engine)

Lets save the table as a csv file

In [9]:
user_comment_df = pd.read_csv("reddit_usernames_comments.csv")

In [10]:
# user_comment_df.to_csv('reddit_usernames_comments.csv', index=False)

In [11]:
sql_for_table2 = """
SELECT
    *
FROM
    public.reddit_usernames;
"""

> Note: The code below may take a while to run. If it fails, reconnect the engine above then rerun the cell

In [12]:
# user_info_df = pd.read_sql_query(sql_for_table2, engine)

In [13]:
user_info_df = pd.read_csv("reddit_usernames.csv")

Lets save the table as a csv file

In [14]:
# user_info_df.to_csv('reddit_usernames.csv', index=False)

Lets take a look at the tables one after the other

In [15]:
user_comment_df.head()

Unnamed: 0              username  \
0           0        LoveAGoodTwist   
1           1            wahznooski   
2           2  Churro_The_fish_Girl   
3           3              abarthch   
4           4            VoodooKing   

                                            comments  
0  Female, Kentucky.  4 years out. Work equine on...  
1  As a woman of reproductive age, fuck Texas|As ...  
2  what makes you want to become a vet?|what make...  
3  I see of course there are changing variables, ...  
4  I have 412+ and faced issues because wireguard...

In [16]:
user_comment_df = user_comment_df.drop(columns="Unnamed: 0")

In [17]:
user_comment_df.head()

username                                           comments
0        LoveAGoodTwist  Female, Kentucky.  4 years out. Work equine on...
1            wahznooski  As a woman of reproductive age, fuck Texas|As ...
2  Churro_The_fish_Girl  what makes you want to become a vet?|what make...
3              abarthch  I see of course there are changing variables, ...
4            VoodooKing  I have 412+ and faced issues because wireguard...

In [18]:
user_comment_df.shape

(3276, 2)

In [19]:
user_info_df.head()

Unnamed: 0             username  isused   subreddit  created_at
0           0       LoveAGoodTwist    True  Veterinary  2024-05-02
1           1            drawntage    True  Veterinary  2024-05-02
2           2           LinkPast84    True  Veterinary  2024-05-02
3           3  heatthequestforfire    True  Veterinary  2024-05-02
4           4       Most-Exit-5507    True  Veterinary  2024-05-02

In [20]:
user_info_df = user_info_df.drop(columns="Unnamed: 0")

In [21]:
user_info_df.head()

username  isused   subreddit  created_at
0       LoveAGoodTwist    True  Veterinary  2024-05-02
1            drawntage    True  Veterinary  2024-05-02
2           LinkPast84    True  Veterinary  2024-05-02
3  heatthequestforfire    True  Veterinary  2024-05-02
4       Most-Exit-5507    True  Veterinary  2024-05-02

In [22]:
user_info_df.shape

(8259, 4)

## Data Exploration

This table (now dataframe) contains usernames of users and their comments

Lets look at a comment in order to understand how it is structured

In [23]:
# print all comments for first user
user_comment_df["comments"][0]

'Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a week Jan-June/July then variable in the off season. No limit on PTO - took ~5 weeks last year. One paid conference a year (registration/travel/ 1/2 hotel/ transportation) or online CE program. All licensures & professional group fees covered. Cell phone allowance and mileage reimbursement.|Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a week Jan-June/July then variable in the off season. No limit on PTO - took ~5 weeks last year. One paid conference a year (registration/travel/ 1/2 hotel/ transportation) or online CE program. All licensures & professional group fees covered. Cell phone allowance and mileage reimbursement.|Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a wee

In [24]:
# split comments into individual comments
first_comments = user_comment_df["comments"][0].split("|")

# get the number of comments for first user
len(first_comments)

16

In [25]:
# remove repeated comments
unique_comment = []
for comment in first_comments:
    if comment in unique_comment:
        continue
    else:
        unique_comment.append(comment)

In [26]:
print(f"Length of unique comments for first user: {len(unique_comment)}")
print()
print(unique_comment)

Length of unique comments for first user: 1

['Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a week Jan-June/July then variable in the off season. No limit on PTO - took ~5 weeks last year. One paid conference a year (registration/travel/ 1/2 hotel/ transportation) or online CE program. All licensures & professional group fees covered. Cell phone allowance and mileage reimbursement.']


It can be seen that the comment column contains multiple comments separated with "|"

It can also be seen that there are repeated comments

Lets check for missing values

In [27]:
user_comment_df.isna().sum()

username    1
comments    0
dtype: int64

In [28]:
user_comment_df[user_comment_df["username"].isna() == True]

username                                           comments
23      NaN  [deleted]|[deleted]|[deleted]|[deleted]|[delet...

In [29]:
user_comment_df.iloc[23]["username"] = "None"

In [30]:
user_comment_df.iloc[23]

username                                                 None
comments    [deleted]|[deleted]|[deleted]|[deleted]|[delet...
Name: 23, dtype: object

In [31]:
user_comment_df.iloc[23]["comments"]

'[deleted]|[deleted]|[deleted]|[deleted]|[deleted]|[deleted]|[deleted]|[removed]|Can I ask a question about really basic vetmed certification? I’m in an area that has a serious shortage of emergency trained vets, so much so that there’s been a pivot to regular vets not doing emergency triage, and not being able to recognize emergencies. \n\nIs there a basic certification that’s available so that pet owners can know when it’s time for the ER?|[deleted]|[deleted]|I agree with some of the below threads. Pay varies from state and I’ve also found big cities tend to pay more than hospitals in burbs or rural areas. For instance, I’m not certified and as a tech in Boston, MA I make $27/hour but in Chicago, IL I made $23/hour. That being said, I live with my boyfriend and having dual incomes is honestly the only way I can afford to live.\n\nI know moving for a job is a big thing consider but maybe not a bad idea to see what’s out there. I’ve also learned to not be afraid to advocate for yoursel

In [32]:
user_comment_df[user_comment_df["username"] == "None"]

username                                           comments
23     None  [deleted]|[deleted]|[deleted]|[deleted]|[delet...

There are no missig values

Let's check if there are duplicate usernames

In [33]:
if user_comment_df["username"].nunique() == user_comment_df.shape[0]:
    print("There are no duplicated usernames")
else:
    print("There are duplicated usernames")

There are no duplicated usernames


Lets explore the second dataframe also

In [34]:
user_info_df.head()

username  isused   subreddit  created_at
0       LoveAGoodTwist    True  Veterinary  2024-05-02
1            drawntage    True  Veterinary  2024-05-02
2           LinkPast84    True  Veterinary  2024-05-02
3  heatthequestforfire    True  Veterinary  2024-05-02
4       Most-Exit-5507    True  Veterinary  2024-05-02

In [35]:
user_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8259 entries, 0 to 8258
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   username    8258 non-null   object
 1   isused      8259 non-null   bool  
 2   subreddit   8259 non-null   object
 3   created_at  8259 non-null   object
dtypes: bool(1), object(3)
memory usage: 201.8+ KB


From the summary above, we se that there are no missing values as each feature has exactly 8259 values which is total entries in the dataset

Let's check if there are duplicate usernames

In [36]:
if user_info_df["username"].nunique() == user_info_df.shape[0]:
    print("There are no duplicated usernames")
else:
    print("There are duplicated usernames")

There are duplicated usernames


In [37]:
user_info_df["username"].nunique()

8258

In [38]:
user_info_df.shape[0]

8259

In [39]:
user_info_df[user_info_df["username"].duplicated() == True]

Empty DataFrame
Columns: [username, isused, subreddit, created_at]
Index: []

At this point lets create a function to preprocess the comments

## Data Preprocessing

Lets define functions to clean the dataset

In [40]:
def remove_web_link(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = re.sub(r"http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\\(\\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+",
                              "", text_list[i].strip())
    return " | ".join(text_list)

In [41]:
def remove_directories(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = re.sub(r"(/[a-zA-Z0-9_]+)+(/)*(.[a-zA-Z_]+)*",
                              "", text_list[i]).strip()
    return " | ".join(text_list)

In [42]:
def remove_deleted_comments(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = re.sub(r"\[deleted\]", "", text_list[i].strip())
    return " | ".join(text_list)

In [43]:
def remove_stopwords(text):
    text_list = text.split("|")
    stop_words = set(stopwords.words('english'))
    for i in range(len(text_list)):
        text_list[i] = " ".join([word for word in text_list[i].split() if word.lower() not in stop_words])
    return " | ".join(text_list)

In [44]:
def remove_punctuations(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = "". join([l if l not in string.punctuation else " " for l in text_list[i]])
        #text_list[i] = ''.join([l for l in text_list[i] if l not in string.punctuation])
    return " | ".join(text_list)

In [45]:
def remove_non_alphabets(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = re.sub(r"[^a-zA-Z ]", "", text_list[i].strip())
    return " | ".join(text_list)

In [46]:
# def autocorrect_spelling(text):
#     spell = Speller()
#     text_list = text.split("|")
#     for i in range(len(text_list)):
#         text_list[i] = spell(text_list[i])
#     return " | ".join(text_list)

In [47]:
def remove_unneeded_spaces(text):
    text_list = text.split("|")
    for i in range(len(text_list)):
        text_list[i] = re.sub(r"(\s)+", " ", text_list[i].strip())
    return " | ".join(text_list)

In [48]:
def remove_repeated_sentence(text):
    text_list = text.split("|")
    unique_comment = []
    for comment in text_list:
        if comment.strip() in unique_comment:
            continue
        else:
            unique_comment.append(comment.strip())
    return " | ".join(unique_comment)

In [49]:
def nlp_preprocessing(text):
    text = remove_web_link(text)
    text = remove_directories(text)
    text = remove_deleted_comments(text)
    text = remove_stopwords(text)
    text = remove_punctuations(text)
    text = remove_non_alphabets(text)
    # text = autocorrect_spelling(text)
    text = remove_unneeded_spaces(text)
    text = remove_repeated_sentence(text)
    text = text.lower()
    return text

## Hand Engineering

Lets check out the unique values in the subreddit feature as well as the count of each value

In [50]:
subreddit_count = user_info_df['subreddit'].value_counts()
subreddit_count

subreddit
Veterinary          6170
MysteriumNetwork     967
medicine             409
HeliumNetwork        400
orchid               303
vet                   10
Name: count, dtype: int64

In [51]:
subreddit_list = list(subreddit_count.index)

Lets explore each of this subreddit categories starting from the least (the bottom)

In [52]:
# get the number of vet subscribers that are in the first dataset

# initialize counter
user_count = 0
# create container for vet subcribers also in the first dataframe
vet_subscribers = []

# for each username who is a subcriber of vet
for user in user_info_df[user_info_df['subreddit'] == "vet"]["username"]:
    # if username is found in table1
    if not user_comment_df[user_comment_df["username"] == user].empty:
        # increment counter by 1
        user_count += 1
        # capture the username
        vet_subscribers.append(user)

print("Vet Subreddit Count")
print("Table1: {}".format(subreddit_count["vet"]))
print(f"Table2: {user_count}")

Vet Subreddit Count
Table1: 10
Table2: 9


One of the subscribers of vet is not in the first dataset

At this point it would be better to combine both dataset into one

Lets do that

In [53]:
reddit_user_df = pd.merge(user_comment_df, user_info_df,
                          on="username", how="left")

In [54]:
reddit_user_df.head()

username                                           comments  \
0        LoveAGoodTwist  Female, Kentucky.  4 years out. Work equine on...   
1            wahznooski  As a woman of reproductive age, fuck Texas|As ...   
2  Churro_The_fish_Girl  what makes you want to become a vet?|what make...   
3              abarthch  I see of course there are changing variables, ...   
4            VoodooKing  I have 412+ and faced issues because wireguard...   

  isused         subreddit  created_at  
0   True        Veterinary  2024-05-02  
1   True        Veterinary  2024-05-02  
2   True        Veterinary  2024-05-02  
3   True  MysteriumNetwork  2024-05-02  
4  False  MysteriumNetwork  2024-05-03

In [55]:
reddit_user_df.iloc[23]["comments"]

'[deleted]|[deleted]|[deleted]|[deleted]|[deleted]|[deleted]|[deleted]|[removed]|Can I ask a question about really basic vetmed certification? I’m in an area that has a serious shortage of emergency trained vets, so much so that there’s been a pivot to regular vets not doing emergency triage, and not being able to recognize emergencies. \n\nIs there a basic certification that’s available so that pet owners can know when it’s time for the ER?|[deleted]|[deleted]|I agree with some of the below threads. Pay varies from state and I’ve also found big cities tend to pay more than hospitals in burbs or rural areas. For instance, I’m not certified and as a tech in Boston, MA I make $27/hour but in Chicago, IL I made $23/hour. That being said, I live with my boyfriend and having dual incomes is honestly the only way I can afford to live.\n\nI know moving for a job is a big thing consider but maybe not a bad idea to see what’s out there. I’ve also learned to not be afraid to advocate for yoursel

In [56]:
reddit_user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3276 entries, 0 to 3275
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   username    3276 non-null   object
 1   comments    3276 non-null   object
 2   isused      3275 non-null   object
 3   subreddit   3275 non-null   object
 4   created_at  3275 non-null   object
dtypes: object(5)
memory usage: 128.1+ KB


In [57]:
reddit_user_df["comments"][0]

'Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a week Jan-June/July then variable in the off season. No limit on PTO - took ~5 weeks last year. One paid conference a year (registration/travel/ 1/2 hotel/ transportation) or online CE program. All licensures & professional group fees covered. Cell phone allowance and mileage reimbursement.|Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a week Jan-June/July then variable in the off season. No limit on PTO - took ~5 weeks last year. One paid conference a year (registration/travel/ 1/2 hotel/ transportation) or online CE program. All licensures & professional group fees covered. Cell phone allowance and mileage reimbursement.|Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a wee

In [58]:
reddit_user_df_processed = reddit_user_df.copy()
reddit_user_df_processed["comments"] = reddit_user_df["comments"].apply(nlp_preprocessing)

In [59]:
reddit_user_df_processed["comments"][0]

'female kentucky years out work equine private practice base salary k plus bonuses k days week jan june limit pto took weeks last year one paid conference year registration transportation online ce program licensures professional group fees covered cell phone allowance mileage reimbursement'

### My Approach to Building the Model

The following are the approaches used to solve this problem
<ol>
    <li>All users would be categorized as others unless proven otherwise from the comments</li>
    <li>Comments are independent of each other (meaning a comment is not continued in another comment)</li>
    <li>When there is indication of user's category in a comment, other comments do not matter (i.e. when users state that they are doctors in a comment, even if other comment are not related to this, the user is still a doctor</li>
    <li>Comments made by a user would be splitted and considered separate data to capture the independence among comments</li>
    <li>When a user is a doctor or a veterinarian, at least one word in the comment that shows the profession must be related to doctor or veterinarian (i.e. when no word in a users comment is realted (similar) to doctor or medicine, automatically the user is not a doctor)</li>
    <li>Any user automatically found from above to not be a doctor or veterinarian would be automatically classified as Others</li>
    <li>The model (to be built) would be built on for only comments having at least a word related to doctor, medicine, veterinarian, animal, hospital or clinic</li>
</ol>

Split comments

In [60]:
user_separated_comment_dict = {
    "username" : [],
    "comment" : [],
    "subreddit" : [],
    "former_index" : []
}

for i in reddit_user_df_processed.index:
    for comment in reddit_user_df_processed.iloc[i]["comments"].split("|"):
        user_separated_comment_dict["username"].append(reddit_user_df_processed.iloc[i]["username"])
        user_separated_comment_dict["comment"].append(comment.strip())
        user_separated_comment_dict["subreddit"].append(reddit_user_df_processed.iloc[i]["subreddit"])
        user_separated_comment_dict["former_index"].append(i)
        
user_separated_comment_df = pd.DataFrame(user_separated_comment_dict)

In [61]:
user_separated_comment_df

username  \
0            LoveAGoodTwist   
1                wahznooski   
2      Churro_The_fish_Girl   
3                  abarthch   
4                  abarthch   
...                     ...   
11185         Real_Use_3216   
11186         Real_Use_3216   
11187         Real_Use_3216   
11188         Real_Use_3216   
11189         Real_Use_3216   

                                                 comment         subreddit  \
0      female kentucky years out work equine private ...        Veterinary   
1                      woman reproductive age fuck texas        Veterinary   
2                                  makes want become vet        Veterinary   
3      see course changing variables dimension change...  MysteriumNetwork   
4      mean far aware people already use torrent prot...  MysteriumNetwork   
...                                                  ...               ...   
11185  earn production everything touch preventatives...        Veterinary   
11186  focus practicing good medicine surgery efficie...        Veterinary   
11187                                               hard        Veterinary   
11188          am crossfit its first thing every workday        Veterinary   
11189                  got whats best move on life short        Veterinary   

       former_index  
0                 0  
1                 1  
2                 2  
3                 3  
4                 3  
...             ...  
11185          3275  
11186          3275  
11187          3275  
11188          3275  
11189          3275  

[11190 rows x 4 columns]

In [62]:
user_separated_comment_df.tail(30)

username  \
11160           daliadeimos   
11161           daliadeimos   
11162           daliadeimos   
11163  Unhappy_Passenger_86   
11164            B1u3Chips_   
11165            B1u3Chips_   
11166           Daktari2018   
11167           Daktari2018   
11168           Daktari2018   
11169           Daktari2018   
11170           Daktari2018   
11171               Sheepb1   
11172               Elyrath   
11173         Real_Use_3216   
11174         Real_Use_3216   
11175         Real_Use_3216   
11176         Real_Use_3216   
11177         Real_Use_3216   
11178         Real_Use_3216   
11179         Real_Use_3216   
11180         Real_Use_3216   
11181         Real_Use_3216   
11182         Real_Use_3216   
11183         Real_Use_3216   
11184         Real_Use_3216   
11185         Real_Use_3216   
11186         Real_Use_3216   
11187         Real_Use_3216   
11188         Real_Use_3216   
11189         Real_Use_3216   

                                                 comment   subreddit  \
11160                                         good point  Veterinary   
11161  clinic work collects payment euths first clien...  Veterinary   
11162  euthanized cat week able void bladder own hes ...  Veterinary   
11163  one also coming difficult situation trying pur...  Veterinary   
11164  im looking applying veterinary nursing college...  Veterinary   
11165  could study college veterinary nursing univers...  Veterinary   
11166  good sticking standards care caring enough spe...  Veterinary   
11167  wonderful wanting know more knowing learn driv...  Veterinary   
11168  its tough come tight group esp area its hard t...  Veterinary   
11169  call company tell length time theyll still gua...  Veterinary   
11170  state youre required pass med clerk exam meds ...  Veterinary   
11171  yes feel free ask someone double check used wo...  Veterinary   
11172  same helps massively errors still occur signfi...  Veterinary   
11173              its different undergrad school school  Veterinary   
11174  mentor ship assigning someone another practice...  Veterinary   
11175  take job knowing make mistakes its part it its...  Veterinary   
11176              field changing change worse im afraid  Veterinary   
11177  best make mistakes learn move life thats point...  Veterinary   
11178  offered something similar essentially mind off...  Veterinary   
11179  one classmates failed first time study second ...  Veterinary   
11180  matter schools say well rounded etc really car...  Veterinary   
11181          loving animals reason become veterinarian  Veterinary   
11182  rough seas make good sailors its part learning...  Veterinary   
11183  im sure ive met practice owner would describe ...  Veterinary   
11184                  whenever decide its time its time  Veterinary   
11185  earn production everything touch preventatives...  Veterinary   
11186  focus practicing good medicine surgery efficie...  Veterinary   
11187                                               hard  Veterinary   
11188          am crossfit its first thing every workday  Veterinary   
11189                  got whats best move on life short  Veterinary   

       former_index  
11160          3269  
11161          3269  
11162          3269  
11163          3270  
11164          3271  
11165          3271  
11166          3272  
11167          3272  
11168          3272  
11169          3272  
11170          3272  
11171          3273  
11172          3274  
11173          3275  
11174          3275  
11175          3275  
11176          3275  
11177          3275  
11178          3275  
11179          3275  
11180          3275  
11181          3275  
11182          3275  
11183          3275  
11184          3275  
11185          3275  
11186          3275  
11187          3275  
11188          3275  
11189          3275

In [63]:
user_separated_comment_df.iloc[11169]["comment"]

'call company tell length time theyll still guarantee vaccine'

Next step is to identify comments where at least a word realated to any of doctor, medicine, veterinarian, animal, hospital or clinic was mentioned

The simiarity index to be used would be cosine similarity and a threshold of 0.7 would be used

To do this, the words would need to be embedded. I would be making use of Glove embedding

First thing is to extract the embedding vectors

In [64]:
embeddings_index = dict()

with open("glove.6B.100d.txt", encoding="utf8") as f:
    for line in f:
        values = line.split()
        word = values[0]
        coefs = np.asarray(values[1:], dtype='float32')
        embeddings_index[word] = coefs

In [65]:
print(f"{len(embeddings_index)} words found")

400000 words found


Next step is to define the vocabulary size

Since we are only concerned with this dataset, the number of unique words in this dataset would form the vocabulary size

In [66]:
words_in_dataset = set()

for comment in user_separated_comment_df["comment"]:
    for word in comment.split():
        words_in_dataset.add(word.lower())

In [67]:
vocabulary_size = len(words_in_dataset)

first_word = list(embeddings_index.keys())[0]
embedding_dim = len(embeddings_index[first_word])

The next step is to tokenize the words

For this the words in Glove embedding would be used as the trainin

In [68]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts(words_in_dataset)

In [69]:
vocabulary_size

18084

The next step is to create embedding matrix

In [70]:
embedding_matrix = np.zeros((vocabulary_size, embedding_dim))

for word, index in tokenizer.word_index.items():
    if index <= vocabulary_size:
        embeddig_vector = embeddings_index.get(word)
        if embeddig_vector is not None:
            embedding_matrix[index-1] = embeddig_vector

In [71]:
print(embedding_matrix.shape)

(18084, 100)


Now lets check the comments containing at least a word that is related to any of doctor, medicine, veterinarian, animal, hospital or clinic as mentioned earlier

In [72]:
base_words = ["doctor", "medicine", "veterinarian", "vet", "animal", "hospital", "clinic", "surgery", "treat"]
base_words_embeddings = []
base_words_magnitudes = []

for base_word in base_words:
    embedding = embeddings_index.get(base_word)
    base_words_embeddings.append(embedding)
    
    magnitude = np.sqrt(np.sum(np.square(embedding)))
    base_words_magnitudes.append(magnitude)

In [73]:
def check_cosine_similarity(text, base_words_embeddings, base_words_magnitudes):
    similarity_list = []
    
    for word in text.split():
        word_embedding = embeddings_index.get(word.lower())
        if np.any(word_embedding):
            word_magnitude = np.sqrt(np.sum(np.square(word_embedding)))

            for i in range(len(base_words_embeddings)):
                similarity = np.dot(word_embedding, base_words_embeddings[i]) / (word_magnitude * base_words_magnitudes[i])
                similarity_list.append(similarity)
    return np.max(similarity_list)

In [74]:
def check_cosine_similarity2(text, base_words_embeddings, base_words_magnitudes):
    similarity_list = []
    word_embeddings = []
    
    for word in text.split():
        word_embedding = embeddings_index.get(word.lower())
        if np.any(word_embedding):
            word_embeddings.append(word_embedding)
    
    if np.any(word_embeddings):
        word_embedding = np.mean(word_embeddings)
        word_magnitude = np.sqrt(np.sum(np.square(word_embedding)))
        for i in range(len(base_words_embeddings)):
            similarity = np.dot(word_embedding, base_words_embeddings[i]) / (word_magnitude * base_words_magnitudes[i])
            similarity_list.append(similarity)
    return np.max(similarity_list)

Checking for similarities does not seem to help

Now, some comments would be labbelled by hand

The approach I would be using for this is to have maximum of 30 comments (or less where comments are not up to 30) from each subreddit to form my training set

Lets take a loo at the number of comments in each subreddit

In [75]:
subreddit_count = user_separated_comment_df['subreddit'].value_counts()
subreddit_count

subreddit
Veterinary          7451
MysteriumNetwork    3506
orchid                28
HeliumNetwork         14
vet                    9
medicine               8
Name: count, dtype: int64

I would be starting from the bottom up to the top

In [76]:
# check all comments in the specified subreddit and label
subreddit = "medicine"
indices = user_separated_comment_df[user_separated_comment_df["subreddit"] == subreddit]["former_index"]
comment_list = reddit_user_df.iloc[indices]["comments"]
for i in indices.values:
    print(f"Comment of user with index {i}")
    print(comment_list[i])
    print()
    print()

Comment of user with index 1438
The elderly man is recovering from hip replacement surgery.


Comment of user with index 1439
The teenage boy was treated for a sports injury.


Comment of user with index 1440
The woman is expecting a baby and visited for a prenatal check-up.


Comment of user with index 1441
I just performed an appendectomy on a patient.


Comment of user with index 1442
The patient’s blood pressure is stabilizing after the medication.


Comment of user with index 1443
The MRI scan revealed a tumor in the patient’s brain.


Comment of user with index 1444
I prescribed antibiotics for the patient’s bacterial infection.


Comment of user with index 1445
The patient’s EKG showed signs of a possible heart attack.




All users in the "medicine" subreddit are medical doctors

It should be noted that these users all have single comments

In [77]:
# check all comments in the specified subreddit and label
subreddit = "vet"
indices = user_separated_comment_df[user_separated_comment_df["subreddit"] == subreddit]["former_index"]
comment_list = reddit_user_df.iloc[indices]["comments"]
for i in indices.values:
    print(f"Comment of user with index {i}")
    print(comment_list[i])
    print()
    print()

Comment of user with index 1446
The puppy was brought in for its first round of vaccinations.


Comment of user with index 1447
The adult horse was treated for laminitis.


Comment of user with index 1448
The juvenile bird was treated for a wing injury.


Comment of user with index 1449
The senior cat was brought in for a routine health check-up.


Comment of user with index 1450
I just performed a neutering procedure on a cat.


Comment of user with index 1451
The dog’s condition is improving after the deworming treatment.


Comment of user with index 1452
The X-ray showed a fracture in the bird’s wing.


Comment of user with index 1453
I prescribed flea prevention medication for the puppy.


Comment of user with index 1454
The horse’s blood test revealed signs of equine infectious anemia.




All users in the "vet" subreddit are Veterinarians

It should be noted that these users all have single comments

In [79]:
# check all comments in the specified subreddit and label
subreddit = "HeliumNetwork"
indices = user_separated_comment_df[user_separated_comment_df["subreddit"] == subreddit]["former_index"].unique()
comment_list = reddit_user_df.iloc[indices]["comments"]
for i in indices:
    print(f"Comment of user with index {i}")
    print(comment_list[i])
    print()
    print()

Comment of user with index 93
I’m getting this too|They just told one of the accounts was compromised. They told to not click in any links|Get rid of windows 😂|I’m getting this too|They just told one of the accounts was compromised. They told to not click in any links|Get rid of windows 😂|I’m getting this too|They just told one of the accounts was compromised. They told to not click in any links|Get rid of windows 😂|I’m getting this too|They just told one of the accounts was compromised. They told to not click in any links|Get rid of windows 😂|I’m getting this too|They just told one of the accounts was compromised. They told to not click in any links|Get rid of windows 😂|I’m getting this too|They just told one of the accounts was compromised. They told to not click in any links|Get rid of windows 😂|I’m getting this too|They just told one of the accounts was compromised. They told to not click in any links|Get rid of windows 😂|I’m getting this too|They just told one of the accounts was 

All users in the "HeliumNetwork" subreddit are neither doctors nor veterinarians

It should be noted that some users have multiple comments and some have repated comments. Reapeted comments would be taken care of during preprocessing

In [80]:
# check all comments in the specified subreddit and label
subreddit = "orchid"
indices = user_separated_comment_df[user_separated_comment_df["subreddit"] == subreddit]["former_index"].unique()
comment_list = reddit_user_df.iloc[indices]["comments"]
for i in indices:
    print(f"Comment of user with index {i}")
    print(comment_list[i])
    print()
    print()

Comment of user with index 205
Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen!|Yassss Queen! | Yassss Queen! ---> /r/MysteriumNetwork/comments/n0es73/the_wait_is_over_mysterium_network_decentralised/gw7eeyk/ | Yea. I am excited about this project too. I am happy to see their [collaboration with Storj](https://mysterium.network/blog/mysterium-and-storj-labs-join-forces/). I am still reviewing their whitepaper and comparing it to the Orchid Protocol. If any one has any cliff's notes I'd appreciate it :). I am interested in using VPNs for increased DAPP security (DDoS attacks). ---> /r/MysteriumNetwork/comments/mrg7bb/just_bought_my_first_myst_tokens_and_feel/guo6ytu/ | My first observation is that Mysterium highlights the intention of splitting up packets to traverse different paths along the VPN network which protects a user from a malicious

All users in the "orchid" subreddit are neither doctors nor veterinarians

It should be noted that some users have multiple comments and some have repated comments. Reapeted comments would be taken care of during preprocessing

The number of users and comments in the remaining two sebreddits are much. Therefore, only 50 comments of users of each category would be added to the training set

In [109]:
random.seed(123)

In [110]:
# randomly select 50 comments made by users in the MysteriumNetwork subreddit
MysteriumNetwork_subreddit_index = user_separated_comment_df[user_separated_comment_df["subreddit"] == "MysteriumNetwork"].index
selected_MysteriumNetwork = random.sample(list(MysteriumNetwork_subreddit_index), 50)

In [111]:
# randomly select 50 comments made by users in the Veterinary subreddit
Veterinary_subreddit_index = user_separated_comment_df[user_separated_comment_df["subreddit"] == "Veterinary"].index
selected_Veterinary = random.sample(list(Veterinary_subreddit_index), 50)

It is time to form my training set

In [146]:
train_indices = []
medicine_indices = list(user_separated_comment_df[user_separated_comment_df["subreddit"] == "medicine"].index)
train_indices.extend(medicine_indices)
vet_indices = list(user_separated_comment_df[user_separated_comment_df["subreddit"] == "vet"].index)
train_indices.extend(vet_indices)
HeliumNetwork_indices = list(user_separated_comment_df[user_separated_comment_df["subreddit"] == "HeliumNetwork"].index)
train_indices.extend(HeliumNetwork_indices)
orchid_indices = list(user_separated_comment_df[user_separated_comment_df["subreddit"] == "orchid"].index)
train_indices.extend(orchid_indices)
train_indices.extend(selected_MysteriumNetwork)
train_indices.extend(selected_Veterinary)

In [147]:
train_set_df = user_separated_comment_df.iloc[train_indices].copy()

In [148]:
train_set_df.loc[:, "Label"] = ""

In [149]:
train_set_df.head()

username                                         comment subreddit  \
5564  test_doctor2  elderly man recovering hip replacement surgery  medicine   
5565  test_doctor3               teenage boy treated sports injury  medicine   
5566  test_doctor4  woman expecting baby visited prenatal check up  medicine   
5567  test_doctor5                  performed appendectomy patient  medicine   
5568  test_doctor6  patients blood pressure stabilizing medication  medicine   

      former_index Label  
5564          1438        
5565          1439        
5566          1440        
5567          1441        
5568          1442

In [150]:
train_set_df.reindex()

username                                            comment  \
5564       test_doctor2     elderly man recovering hip replacement surgery   
5565       test_doctor3                  teenage boy treated sports injury   
5566       test_doctor4     woman expecting baby visited prenatal check up   
5567       test_doctor5                     performed appendectomy patient   
5568       test_doctor6     patients blood pressure stabilizing medication   
...                 ...                                                ...   
9129           daabilge  tbh kind matter picking battles ideally we d p...   
5321          _rosanna_  literally leaving clinic right similar problem...   
4569      extinctplanet  applied internships jobs tab aza com also loca...   
6477  socialdistraction  prices vet care going la its also harder get a...   
6494      Due_Dream9234  technicians assholes you everyone new point ca...   

       subreddit  former_index Label  
5564    medicine          1438        
5565    medicine          1439        
5566    medicine          1440        
5567    medicine          1441        
5568    medicine          1442        
...          ...           ...   ...  
9129  Veterinary          2692        
5321  Veterinary          1345        
4569  Veterinary          1072        
6477  Veterinary          1726        
6494  Veterinary          1738        

[159 rows x 5 columns]

I would label the comments checked above

In [155]:
train_set_df.loc[train_set_df["subreddit"] == "medicine", "Label"] = "Medical Doctor"

In [157]:
train_set_df.loc[train_set_df["subreddit"] == "vet", "Label"] = "Veterinarian"

In [158]:
train_set_df.loc[train_set_df["subreddit"] == "HeliumNetwork", "Label"] = "Other"

In [159]:
train_set_df.loc[train_set_df["subreddit"] == "orchid", "Label"] = "Other"

In [160]:
train_set_df["Label"].value_counts()

Label
                  100
Other              42
Veterinarian        9
Medical Doctor      8
Name: count, dtype: int64

Then, I will now label comments by users in the last two subreddits taking MysteriumNetwork first

In [ ]:
for index in selected_MysteriumNetwork:
    former_index = user_separated_comment_df.iloc[index]["former_index"]
    print(f"Comment of user with index {index}")
    print(user_separated_comment_df.iloc[index]["comment"])
    print()
    print()
    print(reddit_user_df.iloc[former_index]["comments"])
    category = input("Input user category")
    user_separated_comment_df.loc[index, "category"] = category
    print()
    print()
    print("="*20)

Comment of user with index 454
hey sorry ot question safe run node cloud machine i ve oracle free tier i m thinking starting node oracle policies crypto mining strict i m secure it even know myst token mined classic proof work way


Hey, sorry for the OT question, but is it safe run a node on a Cloud machine? I've an oracle free tier and I'm thinking about starting a node here but oracle policies about crypto mining are very strict, so I'm not very secure to do it. (Even if I know that Myst token are not mined in the Classic proof of work way)|Hey, sorry for the OT question, but is it safe run a node on a Cloud machine? I've an oracle free tier and I'm thinking about starting a node here but oracle policies about crypto mining are very strict, so I'm not very secure to do it. (Even if I know that Myst token are not mined in the Classic proof of work way)|Hey, sorry for the OT question, but is it safe run a node on a Cloud machine? I've an oracle free tier and I'm thinking about startin

Input user categoryOther


Comment of user with index 673
node public ip address fine


Use the internal IP address.|Use the internal IP address.|No, you can only host 1 node per public IP address. You can have as many nodes as you wish but again you will need to have that many public IP addresses.|If you at all worried then just enable B2B services only then your node will only transfer business traffic.|Your nodes are working it's a visual bug in the display of monitoring failed.|Yes.|Kryptex handles your payments, I would check with them.|The token have been sent to your selected wallet: [https://polygonscan.com/address/0x6A548987399FAbBbAbb667F59a69A7555F6448CC#tokentxns](https://polygonscan.com/address/0x6A548987399FAbBbAbb667F59a69A7555F6448CC#tokentxns)

&#x200B;

Doesn't look like you have configured Metamask correctly. 

Please note you must first enable your wallet under the Polygon Network and then import custom token MYST from the Polygon Network in order to see the receive

Input user categoryOther


Comment of user with index 5636
use see activity dashboard mysteriumm wireshark


what do you use to see their activity? is there a dashboard on mysteriumm for that or do you Wireshark ---> /r/MysteriumNetwork/comments/xeow68/should_i_change_my_ip_address_every_now_and_then/iojpaab/ | thank you I saw this once and couldn't find it again ---> /r/MysteriumNetwork/comments/x50jy8/as_a_node_runner_what_happens_if_i_format_my_sd/imyhtlt/
Input user categoryOther


Comment of user with index 2514
thats great hope mntd also make thing


We like it fresh and simple Good job!|Thats great! I hope MNTD also make the same thing!!!|We like it fresh and simple Good job!|Thats great! I hope MNTD also make the same thing!!!|We like it fresh and simple Good job!|Thats great! I hope MNTD also make the same thing!!!|We like it fresh and simple Good job!|Thats great! I hope MNTD also make the same thing!!!|We like it fresh and simple Good job!|Thats great! I hope MNTD also make 

Input user categoryOther


Comment of user with index 757
restart system running mysterium try connecting still working check mysterium log say ip port number using


Use the internal IP address.|Use the internal IP address.|No, you can only host 1 node per public IP address. You can have as many nodes as you wish but again you will need to have that many public IP addresses.|If you at all worried then just enable B2B services only then your node will only transfer business traffic.|Your nodes are working it's a visual bug in the display of monitoring failed.|Yes.|Kryptex handles your payments, I would check with them.|The token have been sent to your selected wallet: [https://polygonscan.com/address/0x6A548987399FAbBbAbb667F59a69A7555F6448CC#tokentxns](https://polygonscan.com/address/0x6A548987399FAbBbAbb667F59a69A7555F6448CC#tokentxns)

&#x200B;

Doesn't look like you have configured Metamask correctly. 

Please note you must first enable your wallet under the Polygon Network and the

Input user categoryOther


Comment of user with index 6026
great thanks


Great thanks ---> /r/MysteriumNetwork/comments/s9uwn6/does_mystberry_act_like_a_normal_linux_distro/htr8sm6/ | NetworkChuck actually did a video setting up a node on a headless 32-bit rasbianOS install, I wound up following that to add it to my pi that was already running pi-hole and my other apps. Wound up not using the mystberry image.

[https://www.youtube.com/watch?v=El19X-zHt-c](https://www.youtube.com/watch?v=El19X-zHt-c) ---> /r/MysteriumNetwork/comments/s9uwn6/does_mystberry_act_like_a_normal_linux_distro/humpmle/
Input user categoryOther


Comment of user with index 381
yeah gonna hack right already smaller range


I have 50000 to 60000 udp change it in settings and router. ---> /r/MysteriumNetwork/comments/137fcnb/new_to_this_i_just_wanna_have_reassurance_and/jixryxz/ | If you go to the website where u can view the nice status and all. There are settings which u can adjust accordingly.

U could be conne

Input user categoryOther


Comment of user with index 3154
also ran scprime node started selling licenses node runners me feels bit scammy in compensate lack paying costumers lets generate revenue charging owners node yearly license fee


You can limit traffic to verified business accounts only (used for web/data scraping etc) instead of allowing anyone to use your node. Can be set directly from your node’s dashboard. ---> /r/MysteriumNetwork/comments/12x217l/questions_on_using_and_giving_bandwidth/jhi6w6m/ | Looks like Powershell isn’t installed on your Windows machine? ---> /r/MysteriumNetwork/comments/12ie1uf/i_keep_running_into_this_error/jfte0b4/ | https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.3 ---> /r/MysteriumNetwork/comments/12ie1uf/i_keep_running_into_this_error/jfte354/ | Windows defender flags quite a few VPN clients as malware.

https://www.malwarebytes.com/blog/news/2017/09/explained-false-positives/amp -

Input user categoryOther


Comment of user with index 6129
wont trouble this nothing new tor like dont worry


Yep people were using torrent p2p and no one got hacked… Besides if you are running it on docker which windows users do… so nothing to worry about . 

Still unsure use an old laptop and make it separate… Your funds are safu…

P2P and Decentralisation makes hacking impossible.. Don’t worry. ---> /r/MysteriumNetwork/comments/oqktx6/can_we_trust_mysterium_network_node_provider/h6end5n/ | You won’t be in trouble….This is nothing new..Tor is just like that ..Don’t worry ---> /r/MysteriumNetwork/comments/oqktx6/can_we_trust_mysterium_network_node_provider/h6gcyzr/ | Good thing I am not in west ---> /r/MysteriumNetwork/comments/oqktx6/can_we_trust_mysterium_network_node_provider/helcdc7/ | Good thing  I am from india ---> /r/MysteriumNetwork/comments/oqktx6/can_we_trust_mysterium_network_node_provider/heted87/
Input user categoryOther


Comment of user with index 452
run node hyper v

Input user categoryOther


Comment of user with index 2106
ok people proper agreement isp wondering case really nice isp


Ok so people are doing it 
With out proper agreement from the isp 
I was wondering if that was the case or really nice ISP 😅 ---> /r/MysteriumNetwork/comments/stiz99/best_isp_in_australia_to_run_a_myst_node/hx8cxk2/
Input user categoryOther


Comment of user with index 3282
variable apr


If u jailbreak it u might be able to install that apk|Your fire stick|iOS
This is going to be 👍|YES was waiting for this for a while|If u jailbreak it u might be able to install that apk|Your fire stick|iOS
This is going to be 👍|YES was waiting for this for a while|If u jailbreak it u might be able to install that apk|Your fire stick|iOS
This is going to be 👍|YES was waiting for this for a while|If u jailbreak it u might be able to install that apk|Your fire stick|iOS
This is going to be 👍|YES was waiting for this for a while|If u jailbreak it u might be able to install that apk|Y

Input user categoryOther


Comment of user with index 3937
latest update may see actual github activity yet team working r d researching setting frameworks researching testing required libraries etc result current stage provide us clear knowledge technologies use use order create functioning payments prepare frameworks top work done


From the latest update:

While you may not see any actual GitHub activity yet — that is because our team is working on the R&D, researching and setting up frameworks, researching and testing required libraries, etc... The result of current stage will provide us with a clear knowledge which technologies to use and how to use them in order to create functioning payments and to prepare frameworks on top of all the work that is being done.|From the latest update:

While you may not see any actual GitHub activity yet — that is because our team is working on the R&D, researching and setting up frameworks, researching and testing required libraries, etc... The r

Input user categoryOther


Comment of user with index 2632
well put though want stress dissolve reassemble level privacy definitely goal going happen night previously communicated plan get steps first step deploying initial network nodes testing sorts use cases it discovery connected nodes establishing vpn connection detected nodes payments services provided etc touching deleting posts didnt come us enigma whole team what happened deleted posts one team this believe person wrote comments deleted themselves


I am excited to see what we as a team can deliver in the close/far future. We haven't yet finally planned Q2, but we are looking to work with larger pool of developers and also looking to add CMO and CFO to the team preparing for future developments.
Also looking forward to product advancements as we are getting ready to open the network to 3rd party Service Providers(Node operators) and 3'rd party apps (currently we are getting ready to release our I'st app - "Mysterion")

As for 

In [75]:
user_separated_comment_df["category"] = ""

In [76]:
user_separated_comment_df.head()

username                                            comment  \
0        LoveAGoodTwist  female kentucky years out work equine private ...   
1            wahznooski                  woman reproductive age fuck texas   
2  Churro_The_fish_Girl                              makes want become vet   
3              abarthch  see course changing variables dimension change...   
4              abarthch  mean far aware people already use torrent prot...   

          subreddit  former_index category  
0        Veterinary             0           
1        Veterinary             1           
2        Veterinary             2           
3  MysteriumNetwork             3           
4  MysteriumNetwork             3

In [77]:
subreddit = "Veterinary"
indices = user_separated_comment_df[user_separated_comment_df["subreddit"] == subreddit].index

In [78]:
for index in indices:
    former_index = user_separated_comment_df.iloc[index]["former_index"]
    print(user_separated_comment_df.iloc[index]["comment"])
    print()
    print()
    print(reddit_user_df.iloc[former_index]["comments"])
    category = input("Input user category")
    user_separated_comment_df.iloc[index]["category"] = category
    print()
    print()
    print()

female kentucky years out work equine private practice base salary k plus bonuses k days week jan june limit pto took weeks last year one paid conference year registration transportation online ce program licensures professional group fees covered cell phone allowance mileage reimbursement


Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a week Jan-June/July then variable in the off season. No limit on PTO - took ~5 weeks last year. One paid conference a year (registration/travel/ 1/2 hotel/ transportation) or online CE program. All licensures & professional group fees covered. Cell phone allowance and mileage reimbursement.|Female, Kentucky.  4 years out. Work equine only private practice. Base salary $85k plus bonuses/production which was $20k 2023. 6 days a week Jan-June/July then variable in the off season. No limit on PTO - took ~5 weeks last year. One paid conference a year (registration/trav

KeyboardInterrupt: Interrupted by user

At this point is would be better to build the model on the 4 subreddit we have checked so far then predict the category for the remianing users who subscribe to the remaining 2 categories

## Model Building

For the model building, I would be using glove embedding matrix to embed the words

In [ ]:
# loading glove word vectors (words embeddings) into dictionary
embedding_index = {}

with open('glove.6B.100d.txt', encoding='utf-8') as f:
    for line in f:
        values = line.split()
        word = values[0]
        coefs = np.asarray(values[1:], dtype='float32')
        embedding_index[word] = coefs